crates/typst-ide/src/definition.rs

```diff
@@ -0,0 +1,240 @@
+        }
+        DerefTarget::Ref(r) => {
+            let ref_node = r.cast::<ast::Ref>()?.target();
+            let sel = Selector::Label(Label::new(ref_node));
+            let elem = document?.introspector.query_first(&sel)?;
+            let span = elem.span();
+
+            return Some(Definition {
+                kind: DefinitionKind::Label,
+                name: r.text().clone(),
+                value: elem.label().map(Value::Label),
+                range: span,
+                name_range: Span::detached(),
+            });
+        }
+        DerefTarget::Label(..) | DerefTarget::Normal(..) => {
+            return None;
+        }
+    };
+
+    let mut paths = eco_vec![];
+    while let Some(node) = use_site.cast::<ast::FieldAccess>() {
+        paths.push(node.field().get().clone());
+        use_site = use_site.find(node.target().span())?;
+    }
+
+    let name = use_site.cast::<ast::Ident>()?.get().clone();
+    let src = named_items(world, use_site, |item: NamedItem| {
+        if item.name() != &name {
+            return None;
+        }
+
+        match item {
+            NamedItem::Var(name) => {
+                Some(Definition::item(name.get().clone(), name.span(), None))
+            }
+            NamedItem::Fn(name) => Some(
+                Definition::item(name.get().clone(), name.span(), None)
+                    .with_kind(DefinitionKind::Item(ItemDefinitionKind::Function)),
+            ),
+            NamedItem::Module(item) => Some(Definition::module(item)),
+            NamedItem::Import(name, span, value) => {
+                Some(Definition::item(name.clone(), span, value.cloned()))
+            }
+        }
+    });
+    let src = src.or_else(|| {
+        let in_math = matches!(
+            leaf.parent_kind(),
+            Some(SyntaxKind::Equation)
+                | Some(SyntaxKind::Math)
+                | Some(SyntaxKind::MathFrac)
+                | Some(SyntaxKind::MathAttach)
+        );
+        let library = world.library();
+
+        let scope = if in_math { library.math.scope() } else { library.global.scope() };
+        for (item_name, value) in scope.iter() {
+            if item_name == &name {
+                return Some(Definition::item(
+                    name.clone(),
+                    Span::detached(),
+                    Some(value.clone()),
+                ));
+            }
+        }
+
+        None
+    });
+
+    if paths.is_empty() {
+        return src;
+    }
+
+    None
+}
+
+/// A definition of some item.
+#[derive(Debug, Clone)]
+pub struct Definition {
+    /// The name of the definition.
+    pub name: EcoString,
+    /// The kind of the definition.
+    pub kind: DefinitionKind,
+    /// A possible instance of the definition.
+    pub value: Option<Value>,
+    /// The source range of the entire definition.
+    pub range: Span,
+    /// The range of the name of the definition.
+    pub name_range: Span,
```

Comment:
The name_range seems to always be iehter detached or the same as `range`. Do we need it?

Reply:
The (definition) range is larger or equal to the name range. Providing a range will help editor preview the definition. Providing a name range will help the features like renaming.

I haven't make the definition range yet, since it might introduce much more code.

Should we delete it currently?

![image](https://github.com/typst/typst/assets/35292584/4ef78bf8-85e1-4b42-92f9-959e5a630cf1)

Reply:
Yeah, let's only add it once it's needed.

Reply:
I implemented it.